### PR TITLE
refactor worker_job_heartbeat

### PR DESCRIPTION
**Pull request by zenghsh3**

### CODE REVIEW

File: parl/remote/worker.py [hunk 1]
```diff
@@ -270,26 +270,19 @@ def _init_jobs(self, job_num):
...
-                        "[Worker] lost connection with the job:{}".format(
-                            job.job_address))
-                    if self.master_is_alive and self.worker_is_alive:
-                        self._remove_job(job.job_address)
-
-                return exit_func
-
-            heartbeat_exit_callback_func = get_job_heartbeat_exit_callback_func(
-            )
+            def heartbeat_exit_callback_func(job):
```
Comment (TomorrowIsAnOtherDay) [inline parl/remote/worker.py:273]: It looks more clear now.

File: parl/remote/grpc_heartbeat/heartbeat_client.py [hunk 2]
```diff
@@ -22,22 +22,33 @@
...
+                 exit_func_kwargs={}):
         """Create a thread to run the heartbeat client.
 
             Args:
                 heartbeat_server_addr(str): the address of the heartbeat server.
                 heartbeat_exit_callback_func(function): A callback function, which will be called after the 
-                                                        heartbeat exit. There should be no arguments in the 
-                                                        function.
+                                                        heartbeat exit.
+                exit_func_args(tuple): the argument tuple for the heartbeat_exit_callback_func invocation. Defaults to ().
```
Comment (TomorrowIsAnOtherDay) [inline parl/remote/grpc_heartbeat/heartbeat_client.py]: calling the heartbeat_exit_callback_func.

File: parl/remote/grpc_heartbeat/heartbeat_client.py [hunk 3]
```diff
@@ -22,22 +22,33 @@
...
+                exit_func_kwargs(dict): the argument tuple for the heartbeat_exit_callback_func invocation. Defaults to {}.
```
Comment (TomorrowIsAnOtherDay) [inline parl/remote/grpc_heartbeat/heartbeat_client.py]: the argument `dict` for calling the heartbeat_exit_callback_func.

File: parl/remote/grpc_heartbeat/tests/heartbeat_server_arguments_test.py [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+        heartbeat_client_thread.start()
+
+        time.sleep(remote_constants.HEARTBEAT_RCVTIMEO_S * 2)
+
+        # check server and client are still alive after HEARTBEAT_RCVTIMEO_S * 2
+        assert heartbeat_server_thread.is_alive()
+        assert heartbeat_client_thread.is_alive()
+
+        heartbeat_server_thread.exit()  # manually exit the server
+        # will raise an exception in the backend thread
```
Comment (TomorrowIsAnOtherDay) [inline parl/remote/grpc_heartbeat/tests/heartbeat_server_arguments_test.py:189]: This is a good test, but the main thread should know that something has gone wrong when calling the exit function.

File: parl/remote/grpc_heartbeat/heartbeat_server.py [hunk 5]
```diff
@@ -72,16 +72,22 @@ def timeout_timer(self):
...
+                 exit_func_args=(),
+                 exit_func_kwargs={}):
         """Create a thread to run the heartbeat server.
 
             Args:
                 heartbeat_exit_callback_func(function): A callback function, which will be called after the 
-                                                        heartbeat exit. There should be no arguments in the 
-                                                        function.
+                                                        heartbeat exit.
+                exit_func_args(tuple): the argument tuple for the heartbeat_exit_callback_func invocation. Defaults to ().
```
Comment (TomorrowIsAnOtherDay) [inline parl/remote/grpc_heartbeat/heartbeat_server.py]: Please update the comments for these two lines, following the same suggestions at the `hearbeat_client.py`.

File: parl/remote/grpc_heartbeat/heartbeat_client.py (hunk 3 above)
Comment (zenghsh3) [reply parl/remote/grpc_heartbeat/heartbeat_client.py]: fixed

File: parl/remote/grpc_heartbeat/heartbeat_server.py (hunk 5 above)
Comment (zenghsh3) [reply parl/remote/grpc_heartbeat/heartbeat_server.py]: fixed

File: parl/remote/grpc_heartbeat/heartbeat_client.py (hunk 2 above)
Comment (zenghsh3) [reply parl/remote/grpc_heartbeat/heartbeat_client.py]: fixed

**Review by TomorrowIsAnOtherDay — APPROVED** [review]
LGTM.

**Pull request merged**